### ISSUE TITLE
whether BN operation can be done in-place

### ISSUE THREAD

**Issue by animebing** [yjxiong/caffe#165]
@yjxiong This is the subsequent issue of #163 ,in your last answer, you say the BN operation cannot be done in-place no matter in CAFFE engine or CUDNN engine, I think your answer is partly correct.
(I am sorry that the link below cannot go to the right page, I can't figure out why, so you need to go there by hand)
1. you are correct that in CUDNN engine, BN cannot be done in-place, because in the backward computation, the bottom_data is used, if this is be done in-place, what we get here is the top_data, which is absolutely uncorrect, so BN operation cannot be done in-place in CUDNN engine.
2. In CAFFE engine, if the BN operation is done in-place(CPU or GPU), for the bottom_data and top_data, in the Forward_gpu, the value of bottom_data is changed in [https://github.com/yjxiong/caffe/blob/action_recog/src/caffe/layers/bn_layer.cu#L52](url)  because top_data is written into data, then bottom_data is not used any more in the forward computation, at the same time, in the Backward_gpu, the bottom_data is not used, so it is okay for the bottom_data and top_data to use the same memory. Then for the bottom_diff and top_diff, in the Backward_gpu, the top_diff is used twice to compute the diff for slope and bias, it is okay, then in [https://github.com/yjxiong/caffe/blob/action_recog/src/caffe/layers/bn_layer.cu#L202](url), the value of top_diff is changed because bottom_diff is written into data, but from here, the top_diff is not used any more, so it is okay for the bottom_diff and top_diff to use the same memory. Based on above, it is okay to do BN operation in-place in CAFFE engine.

Please tell me if my analysis is correct.

**Comment by yjxiong** [comment]
It is true that the current implement of `BN` can run under the in-place mode. But it relies on the fact that `x_norm` and `x_inv_std` are saved during forward. See here

https://github.com/yjxiong/caffe/blob/action_recog/src/caffe/layers/bn_layer.cu#L96

This may not be true in future. So in-place operation is not safe to future changes.

In `SyncBN`, where we used another implementation, you will get incorrect results if running in in-place mode. This is confirmed by our previous experiments.

Besides, using in-place operation does not save much memory space, especially in testing, where memory for blobs are automatically shared to produce minimal footprint.

To sum up, it is pointless to perform in-place operation for BN related layers. It's dangerous, unsupported and not benefitial. But surely it is your call if your really like it that way.

**Comment by animebing** [comment]
@yjxiong Thanks for your answer, which helps me a lot
